classes/submissionFile/Repository.inc.php

```diff
@@ -722,7 +722,7 @@ public function getWorkflowStageId(SubmissionFile $submissionFile): ?int
             $query = $queryDao->getById($note->getAssocId());
 
             // The note should be associated with a query. If not, fail.
-            if ($query->getAssocType() != PKPApplication::ASSOC_TYPE_QUERY) {
+            if ($query?->getAssocType() != PKPApplication::ASSOC_TYPE_QUERY) {
```

Comment:
@asmecher it seems to me like this is a data inconsistency (`SubmissionFile` associated with a `Note` that exists without a `Query`). Should we be looking into this further or just moving on?

Reply:
Agreed, and unfortunately this isn't something that'll be fixed with the addition of foreign keys, since it's using the assoc type/ID pairing. @defstat, can you investigate how this data came to be?

Reply:
@asmecher @NateWr I reproduced it by using the following sequence of actions:

1. Go to a submission's workflow page
2. Go to the Copyediting workflow tab (for example)
3. On the "Copyediting Discussions" grid, hit on "Add discussion"
4. On the Query form that comes up, Attach a new file using "Upload File"
5. Cancel the form

Those sequence of events, up until the step 5, in the DB 3 objects, connected to each other, are created:
- One Query in the `Query` table.
- One Note in the `Notes` table.
- One SubmissionFile in the `SubmissionFiles` table.

After performing step 5 (the cancel form action) of the above sequence, the `deleteQuery` function is executed (https://github.com/pkp/pkp-lib/blob/main/controllers/grid/queries/QueriesGridHandler.inc.php#L362). That function seems to delete the query from the database, but not all the other related objects (the SubmissionFile and the Note). The query is deleted here (https://github.com/pkp/pkp-lib/blob/main/controllers/grid/queries/QueriesGridHandler.inc.php#L370).

I can see 3 ways of action here:
1. Incorporate the removal of all related to the query notes in the `queryDAO -> deleteObject`, and incorporate the removal of all Note related SubmissionFiles for the deleted Note in the `noteDAO -> deleteObject`
2. Incorporate the removal of all related objects in the `deleteQuery` function (https://github.com/pkp/pkp-lib/blob/main/controllers/grid/queries/QueriesGridHandler.inc.php#L362) (with the danger to have another code path that will possibly produce the same problem)
3. Close this issue as is (with the nullable fix) and wait for the repository pattern development for the Query entity, and incorporate the change to the Query repository (which I think will be very similar to 2 above in practice).

I think I would go for the 1. above as the best solution.

Reply:
@asmecher @NateWr  I pushed a try to rewrite the `deleteById` functions by using transactions. Could you please check if that's in the correct direction?

Reply:
This particular line causes another issue, reported by a user on the forum: https://forum.pkp.sfu.ca/t/ojs-journal-is-having-an-issue-with-uploading-documents/73718 - authorization denial to upload a file on the discussion. This is because in this case query object has assoc_id `ASSOC_TYPE_SUBMISSION` and stage id cannot be determined for authorization purposes. I've made a comment on the original issue: https://github.com/pkp/pkp-lib/issues/7563#issuecomment-1177307890

Reply:
@Vitaliy-1 you mean the line without the fix (this one `if ($query->getAssocType() != PKPApplication::ASSOC_TYPE_QUERY) {`) causes the issue you mention? Or has someone tried the fix (this one `if ($query?->getAssocType() != PKPApplication::ASSOC_TYPE_QUERY) {` and has encountered the mentioned issue?

Reply:
@Vitaliy-1 By reading the comments/issue/forum again, I understand that the fix (`if ($query?->getAssocType() != PKPApplication::ASSOC_TYPE_QUERY) {`) is causing the [forum's ](https://forum.pkp.sfu.ca/t/ojs-journal-is-having-an-issue-with-uploading-documents/73718) issue. 

At least for the Journal Manager, I don't seem to be able to reproduce the mentioned error though, when trying to upload a file (it is uploaded as it supposed to). Is that because there is a fix for that on the main branch (you mention something like that [here](https://github.com/pkp/pkp-lib/issues/7563#issuecomment-1177307890)) or am I missing something?

Reply:
I can reproduce this problem described on the forum on the current main branch using the user with `sub_editor` role. And yes, the user applied the patch on the stable branch. This patch was applied to our main branch but not stable

Reply:
It's about original line, not your particular change. `$query->getAssocType()` returns `ASSOC_TYPE_SUBMISSION` in that particular case and fails the check to determine workflow stage id. Without this ID, one of the authorization policies, in particular `SubmissionFileMatchesWorkflowStageIdPolicy`, always fails.

Reply:
@defstat, about the use of transactions: I don't think it's necessary in this case, and since nested transactions are not well supported in e.g. MySQL, I'd like to save the use of transactions until we have a good use case at a higher level. I think just deleting the dependent elements in reverse sequence of creation (like you've done) is good enough here.

Reply:
@asmecher @NateWr 

- Removed transactions (left the rewrite of the functions as is)
- Added the data/file Migrations (separate commit). Here I am not sure whether I can use the following 
     - `Application::ASSOC_TYPE_QUERY` (At [this](https://github.com/pkp/pkp-lib/blob/main/classes/migration/upgrade/PKPv3_3_0UpgradeMigration.inc.php) the value is copied like [such](https://github.com/pkp/pkp-lib/blob/main/classes/migration/upgrade/PKPv3_3_0UpgradeMigration.inc.php#L28))
     - `SubmissionFile::SUBMISSION_FILE_QUERY` (At [this](https://github.com/pkp/pkp-lib/blob/main/classes/migration/upgrade/PKPv3_3_0UpgradeMigration.inc.php) the value is copied like [such](https://github.com/pkp/pkp-lib/blob/main/classes/migration/upgrade/PKPv3_3_0UpgradeMigration.inc.php#L43))
     - `Services::get('file');` (At [this](https://github.com/pkp/pkp-lib/blob/main/classes/migration/upgrade/PKPv3_3_0UpgradeMigration.inc.php) the service's adapter is used like [such](https://github.com/pkp/pkp-lib/blob/main/classes/migration/upgrade/PKPv3_3_0UpgradeMigration.inc.php#L424))

Reply:
Looks good, @defstat! Please do copy the constant values and put the constant name in a comment as the other examples you've linked have done. This will be useful if the constants ever change in the future -- we might do this to reconcile the sometimes different constants used in different applications -- e.g. OJS uses `0x100` for `ASSOC_TYPE_JOURNAL` and OMP uses `0x200` for `ASSOC_TYPE_PRESS`. If we later amalgamate these into one and change the values in one app, it'll be important that the migration classes don't break when they encounter old data.

Reply:
@asmecher thanks. I made the changed, pushed, and waiting for the tests to finish

Reply:
I'm receiving the same error with permission restriction to file upload to the discussion. To reproduce on the test instance:
1. Go to the submission workflow, e.g., with ID 6 - Philips...
2. Login as editor assigned to the article, Daniel Barnes
3. Open a production dicussion by clicking on `Add Discussion`, then click on `Upload File`.
4. For debugging purposes, e.g., print `$query->getAssocType()` to the log before this line.
5. Upload a file and click `continue`
6.  See the message `The current role does not have access to this operation.`
7. See the error log, where query assoc type is `1048585` (`ASSOC_TYPE_SUBMISSION`) and not `1048586` (`ASSOC_TYPE_QUERY`)

Reply:
@defstat, Would you like me to open a new issue for this one or you are planning to handle it here?

Reply:
@Vitaliy-1 this can be reproduced on both `main` branch and this (#8073) branch. I now understand what you meant when you said that the changes for this particular issue are not causing the `The current role does not have access to this operation.` problem, but they are not fixing it either. 

Not sure if everyone agree, but I think it's an entirely different problem than this one, and should be dealt in a separate issue, like you are proposing [here](https://github.com/pkp/pkp-lib/pull/8075#discussion_r921025381).

Reply:
I'm mentioning it here because these lines were introduced in the main branch to fix the fatal error when opening the activity log. See: https://github.com/pkp/pkp-lib/issues/7563. I don't think them have any other rationale, so I'm wondering if it's now as simple as deleting those checks

Reply:
That seems to be a business logic question that I can't answer, not feeling very familiar with the exact business logic behind the relationships between fileStages and Notes/Queries AssocTypes.

Is having a `query` there, that has an `assocType` other than `PKPApplication::ASSOC_TYPE_QUERY`, consistent business-wise? 

In the case of @Vitaliy-1 's reproduction [steps](https://github.com/pkp/pkp-lib/pull/8075#discussion_r921024848), the `$query->getAssocType()` has the value of `PKPApplication::ASSOC_TYPE_SUBMISSION`. Is this acceptable business-wise?

Another thing to consider here is that when I tried to reproduce the same problem using a clean database, I did not reproduce it, but using the ojs test data (I took them from pkp/datasets), I did reproduce it. 

@asmecher, @NateWr  any insights on that?

Reply:
I am wondering whether there is a case where a Query has an `AssocType` different than `PKPApplication::ASSOC_TYPE_SUBMISSION` and if so in which cases.

Reply:
I'm not aware of a case where a query has a different `assoc_type`. However, I believe it is possible that a `note` exists without a `query`. I think a `note` can be added on a submission file, and it uses the same `notes` database that the query notes use.

To create a note on a submission file:

1. Upload a submission file to submission files, review files, revisions, etc.
2. Open up "More Information" on the submission file.
3. Open "Notes".
4. Add a "note".
5. See the new note in the `notes` table.

Reply:
@NateWr that's correct. 

1. This note does not come with a submission file though, meaning that, at least the way the code is now, the addition of this kind of note will not go through the code segment that the change refers to.
2. The migration selects the notes that have no queries but have the `assoc_type = ASSOC_TYPE_QUERY`, which is not the assoc_type of a note from the process @NateWr describes (which is `ASSOC_TYPE_SUBMISSION_FILE`)
3. There is yet another process that produces notes without queries. To create it you can do the following
     - On submission workflow go to Activity Log
     - Click on "Notes" Tab
     - Add a note.
     - The note is added without a query with `assoc_type=ASSOC_TYPE_SUBMISSION`

I think that @Vitaliy-1 is right - the if clause makes no sense there as far as I can understand the query states. Not sure why it is there though. We may have missed something. Does anyone agree that this check there can be removed?(@asmecher @NateWr @jonasraoni) (I will remove it and run the tests to see what is happening)

Reply:
It's entirely possible that a `submissionFile\Repository` method did not take into account that a note might not be used in discussions. It's totally ok to adjust the method for that.

However, we should be careful to look at what this method is doing. `submissionFile/Repository::getWorkflowStageId()` should only ever concern itself with a submission file. If that method is being called by something that isn't related to a submission file, it's the calling code that should be changed.

In other words, a note added through the activity log should never encounter this method.

Reply:
I think this all stems back to a change that got introduced in https://github.com/pkp/pkp-lib/commit/129da5afb974dd47c5e44ce4c52700addb1a099f. The following should fix it:
```diff
diff --git a/classes/submissionFile/Repository.inc.php b/classes/submissionFile/Repository.inc.php
index 24fd2496a3..05fe48db1a 100644
--- a/classes/submissionFile/Repository.inc.php
+++ b/classes/submissionFile/Repository.inc.php
@@ -722,7 +722,7 @@ abstract class Repository
             $query = $queryDao->getById($note->getAssocId());
 
             // The note should be associated with a query. If not, fail.
-            if ($query->getAssocType() != PKPApplication::ASSOC_TYPE_QUERY) {
+            if ($note->getAssocType() != PKPApplication::ASSOC_TYPE_QUERY) {
                 return null;
             }
 ```

Reply:
Good find! In the original code, this check was _before_ `$queryDao->getById($note->getAssocId())` is called:

https://github.com/pkp/pkp-lib/commit/129da5afb974dd47c5e44ce4c52700addb1a099f#diff-007b39807ba72de25a23e46d4ac9235dc5328d76fb97f461b45a47061785495dL809-L816

Reply:
I re-added the initial code, with the check on the `note` and not on the `query`. I rebased and pushed the change.